consensus/src/lib.rs

```diff
@@ -174,5 +174,23 @@ cfg_if::cfg_if! {
             /// may change (and old shares may be deleted).
             fn share(&self, index: Self::Index) -> Option<&Self::Share>;
         }
+
+        /// Monitor is the interface an external actor can use to observe the progress of a consensus implementation.
+        ///
+        /// Monitor is used to implement mechanisms that share the same set of active participants as consensus and/or
+        /// perform some activity that requires some synchronization with the progress of consensus.
+        ///
+        /// Monitor can be implemented using [`Committer`](crate::Committer) to avoid introducing complexity
+        /// into any particular consensus implementation.
+        pub trait Monitor: Clone + Send + 'static {
+            /// Index is the type used to indicate the in-progress consensus decision.
+            type Index;
+
+            /// Latest index known by the consensus implementation.
+            fn latest(&self) -> Self::Index;
+
+            /// Create a channel that will receive updates when the latest index changes.
+            fn subscribe(&mut self) -> mpsc::Receiver<Self::Index>;
```

Comment:
To avoid a race condition during async, should `subscribe` simply return a tuple `(latest, receiver)`?

Otherwise you could either:
- Call `latest` then `subscribe`, possibly dropping an update that occurs between the two calls
- Call `subscribe` then `latest`, possibly processing a message from the subscription that updates to a value less than the `latest` value

I think (?) this can happen in a concurrent environment even though the two functions are not `async`

Reply:
Ah interesting. That's probably a good idea.